expertsystem/state/conservation_rules.py

```diff
@@ -634,106 +429,145 @@
+    l_magnitude: NodeQuantumNumbers.l_magnitude
+    l_projection: NodeQuantumNumbers.l_projection
+    s_magnitude: NodeQuantumNumbers.s_magnitude
+    s_projection: NodeQuantumNumbers.s_projection
+
+
+def _check_spin_couplings(
+    in_part: List[Spin],
+    out_part: List[Spin],
+    interaction_qns: Optional[SpinNodeInput],
+) -> bool:
+    in_tot_spins = __calculate_total_spins(in_part, interaction_qns)
+    out_tot_spins = __calculate_total_spins(out_part, interaction_qns)
     matching_spins = in_tot_spins.intersection(out_tot_spins)
     if len(matching_spins) > 0:
         return True
     return False
 
 
-def __calculate_total_spins(part_list, interaction_qns, use_projection):
-    # pylint: disable=too-many-branches
-    total_spins = set()
-    if len(part_list) == 1:
-        if use_projection:
-            total_spins.add(part_list[0])
-        else:
-            spin_magnitude = part_list[0].magnitude
-            total_spins.add(Spin(spin_magnitude, spin_magnitude))
-    else:
-        # first couple all spins together
-        spins_daughters_coupled = set()
-        spin_list = deepcopy(part_list)
-        while spin_list:
-            if spins_daughters_coupled:
-                temp_coupled_spins = set()
-                tempspin = spin_list.pop()
-                for spin in spins_daughters_coupled:
-                    coupled_spins = __spin_couplings(
-                        spin, tempspin, use_projection
-                    )
-                    temp_coupled_spins.update(coupled_spins)
-                spins_daughters_coupled = temp_coupled_spins
-            else:
-                spins_daughters_coupled.add(spin_list.pop())
-        if InteractionQuantumNumberNames.L in interaction_qns:
-            ang_mom = interaction_qns[InteractionQuantumNumberNames.L]
-            spin = interaction_qns[InteractionQuantumNumberNames.S]
-            if use_projection:
-                if spin in spins_daughters_coupled:
-                    total_spins.update(
-                        __spin_couplings(spin, ang_mom, use_projection)
-                    )
-            else:
-                if spin.magnitude in [
-                    x.magnitude for x in spins_daughters_coupled
-                ]:
-                    total_spins.update(
-                        __spin_couplings(spin, ang_mom, use_projection)
+def _check_magnitude(
+    in_part: List[float],
+    out_part: List[float],
+    interaction_qns: Optional[SpinNodeInput],
+) -> bool:
+    def couple_mags(j_1: float, j_2: float) -> List[float]:
+        return [
+            x / 2.0
+            for x in range(
+                int(2 * abs(j_1 - j_2)), int(2 * (j_1 + j_2 + 1)), 2
+            )
+        ]
+
+    def couple_magnitudes(
+        magnitudes: List[float], interaction_qns: Optional[SpinNodeInput]
+    ) -> Set[float]:
+        if len(magnitudes) == 1:
+            return set(magnitudes)
+
+        coupled_magnitudes = set([magnitudes[0],])
```

Comment:
Why is a `,` needed in `[magnitudes[0],]`? That would be same as `[magnitudes[0], None]`, right?

Reply:
I believe it removes the list so it would be a `Set[float]` instead of `Set[List[float]]`

Reply:
This is just a runtime cast I think, not anything about type hints

Reply:
You are right, i just tried that

Reply:
Probably just a consequence of `black`. This behaviour (namely, keeping the `,` after removing the line break) is fixed with version 20 (see #252).